File: ios/Sources/ILDSDesignSystem/IldsToast.swift
import ILDSTokens
import SwiftUI

// Figma set 17708:3491 — mirrors web/Toast.tsx + android/IldsToast.kt.

public enum IldsToastVariant: Sendable {
    case info, success, warning, error
}

public struct IldsToastAction {
    public let label: String
    public let action: () -> Void

    public init(label: String, action: @escaping () -> Void) {
        self.label = label
        self.action = action
    }
}

public struct IldsToastActions {
    public let primary: IldsToastAction?
    public let secondary: IldsToastAction?

    public init(primary: IldsToastAction? = nil, secondary: IldsToastAction? = nil) {
        self.primary = primary
        self.secondary = secondary
    }
}

public struct IldsToast: View {
    private let message: String
    private let heading: String?
    private let variant: IldsToastVariant
    private let showClose: Bool
    private let actions: IldsToastActions?
    private let onClose: (() -> Void)?

    public init(
        variant: IldsToastVariant,
        message: String,
        heading: String? = nil,
        showClose: Bool = false,
        actions: IldsToastActions? = nil,
        onClose: (() -> Void)? = nil
    ) {
        self.variant = variant
        self.message = message
        self.heading = heading
        self.showClose = showClose
        self.actions = actions
        self.onClose = onClose
    }

    public var body: some View {
        let style = IldsToastStyle.resolve(variant)
        let hasHeading = heading.map { !$0.isEmpty } ?? false

        VStack(alignment: .leading, spacing: ILDSTokens.sp12) {
            HStack(alignment: .top, spacing: ILDSTokens.sp8) {
                Image(systemName: style.iconName)
                    .font(.system(size: 24))
                    .foregroundStyle(style.iconColor)
                    .frame(width: 24, height: 24)

                VStack(alignment: .leading, spacing: ILDSTokens.sp4) {
                    if hasHeading, let heading {
                        Text(heading)
                            .font(.system(size: ILDSTokens.fontSize14, weight: ILDSTokens.fontWeightBold))
                            .foregroundStyle(ILDSTokens.neutralCoolgray900)
                    }
                    Text(message)
                        .font(.system(size: ILDSTokens.fontSize14, weight: ILDSTokens.fontWeightRegular))
                        .foregroundStyle(ILDSTokens.neutralCoolgray800)
                        .fixedSize(horizontal: false, vertical: true)
                }
                .frame(maxWidth: .infinity, alignment: .leading)

                if showClose, onClose != nil {
                    Button(action: { onClose?() }) {
                        Image(systemName: "xmark")
                            .font(.system(size: ILDSTokens.sp20))
                            .foregroundStyle(ILDSTokens.neutralCoolgray500)
                    }
                    .buttonStyle(.plain)
                    .accessibilityLabel("Close notification")
                }
            }

            if actions?.secondary != nil || actions?.primary != nil {
                HStack(spacing: ILDSTokens.sp8) {
                    if let secondary = actions?.secondary {
                        IldsButton(
                            secondary.label,
                            action: secondary.action,
                            type: .secondary,
                            size: .medium
                        )
                    }
                    if let primary = actions?.primary {
                        IldsButton(
                            primary.label,
                            action: primary.action,
                            type: .primary,
                            size: .medium
                        )
                    }
                }
            }
        }
        .padding(ILDSTokens.sp12)
        .frame(maxWidth: 320)
        .background(ILDSTokens.globalWhite000)
        .overlay {
            RoundedRectangle(cornerRadius: ILDSTokens.radiusXlarge)
                .strokeBorder(style.border, lineWidth: 1)
        }
        .clipShape(RoundedRectangle(cornerRadius: ILDSTokens.radiusXlarge))
        .shadow(color: ILDSTokens.neutralCoolgray300.opacity(0.5), radius: 6, x: 0, y: 4)
        .accessibilityElement(children: .combine)
        .accessibilityLabel(hasHeading ? "\(heading ?? ""). \(message)" : message)
    }
}

private struct IldsToastStyle {
    let border: Color
    let iconColor: Color
    let iconName: String

    static func resolve(_ variant: IldsToastVariant) -> IldsToastStyle {
        switch variant {
        case .success:
            return IldsToastStyle(
                border: ILDSTokens.successGreen50,
                iconColor: ILDSTokens.successGreen500,
                iconName: "checkmark.circle"
            )
        case .info:
            return IldsToastStyle(
                border: ILDSTokens.secondaryBlue50,
                iconColor: ILDSTokens.informativeBlue500,
                iconName: "info.circle"
            )
        case .warning:
            return IldsToastStyle(
                border: ILDSTokens.warningAmber50,
                iconColor: ILDSTokens.warningAmber500,
                iconName: "exclamationmark.triangle"
            )
        case .error:
            return IldsToastStyle(
                border: ILDSTokens.errorRed50,
                iconColor: ILDSTokens.errorRed600,
                iconName: "exclamationmark.circle"
            )
        }
    }
}

File: android/ilds-design-system/src/main/kotlin/com/icicilombard/ilds/components/IldsToast.kt
package com.icicilombard.ilds.components

import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.CheckCircle
import androidx.compose.material.icons.filled.Close
import androidx.compose.material.icons.filled.Error
import androidx.compose.material.icons.filled.Info
import androidx.compose.material.icons.filled.Warning
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.Immutable
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.semantics.LiveRegionMode
import androidx.compose.ui.semantics.liveRegion
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.icicilombard.ilds.tokens.IldsTokens

// Figma set 17708:3491 — mirrors web/Toast.tsx + lib/ilds_toast.dart.

enum class IldsToastVariant { Success, Info, Warning, Error }

data class IldsToastAction(
    val label: String,
    val onClick: () -> Unit,
)

data class IldsToastActions(
    val primary: IldsToastAction? = null,
    val secondary: IldsToastAction? = null,
)

@Composable
fun IldsToast(
    variant: IldsToastVariant,
    message: String,
    modifier: Modifier = Modifier,
    heading: String? = null,
    showClose: Boolean = false,
    actions: IldsToastActions? = null,
    onClose: (() -> Unit)? = null,
) {
    val style = remember(variant) { IldsToastStyle.resolve(variant) }
    val liveRegionMode = if (variant == IldsToastVariant.Success || variant == IldsToastVariant.Info) {
        LiveRegionMode.Polite
    } else {
        LiveRegionMode.Assertive
    }

    Surface(
        modifier = modifier
            .fillMaxWidth()
            .semantics { liveRegion = liveRegionMode },
        shape = RoundedCornerShape(IldsTokens.radiusXlarge),
        color = IldsTokens.globalWhite000,
        shadowElevation = 8.dp,
        border = BorderStroke(1.dp, style.border),
    ) {
        Column(
            modifier = Modifier.padding(IldsTokens.sp12),
            verticalArrangement = Arrangement.spacedBy(IldsTokens.sp12),
        ) {
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.spacedBy(IldsTokens.sp8),
                verticalAlignment = Alignment.Top,
            ) {
                Icon(
                    imageVector = style.icon,
                    contentDescription = null,
                    tint = style.iconColor,
                    modifier = Modifier.size(24.dp),
                )
                Column(
                    modifier = Modifier.weight(1f),
                    verticalArrangement = Arrangement.spacedBy(IldsTokens.sp4),
                ) {
                    if (heading != null) {
                        Text(
                            text = heading,
                            style = TextStyle(
                                fontSize = IldsTokens.fontSize14,
                                lineHeight = 18.sp,
                                fontWeight = IldsTokens.fontWeightBold,
                            ),
                            color = IldsTokens.neutralCoolgray900,
                        )
                    }
                    Text(
                        text = message,
                        style = TextStyle(
                            fontSize = IldsTokens.fontSize14,
                            lineHeight = 18.sp,
                            fontWeight = IldsTokens.fontWeightRegular,
                        ),
                        color = IldsTokens.neutralCoolgray800,
                    )
                }
                if (showClose && onClose != null) {
                    IconButton(
                        onClick = onClose,
                        modifier = Modifier.size(20.dp),
                    ) {
                        Icon(
                            imageVector = Icons.Filled.Close,
                            contentDescription = "Close notification",
                            tint = IldsTokens.neutralCoolgray500,
                            modifier = Modifier.size(20.dp),
                        )
                    }
                }
            }
            val hasActions = actions?.primary != null || actions?.secondary != null
            if (hasActions) {
                Row(horizontalArrangement = Arrangement.spacedBy(IldsTokens.sp8)) {
                    actions?.secondary?.let { action ->
                        IldsButton(
                            label = action.label,
                            onClick = action.onClick,
                            type = IldsButtonType.Secondary,
                            size = IldsButtonSize.Medium,
                        )
                    }
                    actions?.primary?.let { action ->
                        IldsButton(
                            label = action.label,
                            onClick = action.onClick,
                            type = IldsButtonType.Primary,
                            size = IldsButtonSize.Medium,
                        )
                    }
                }
            }
        }
    }
}

@Immutable
private data class IldsToastStyle(
    val border: Color,
    val iconColor: Color,
    val icon: ImageVector,
) {
    companion object {
        fun resolve(variant: IldsToastVariant): IldsToastStyle = when (variant) {
            IldsToastVariant.Success -> IldsToastStyle(
                IldsTokens.successGreen50,
                IldsTokens.successGreen500,
                Icons.Filled.CheckCircle,
            )
            IldsToastVariant.Info -> IldsToastStyle(
                IldsTokens.secondaryBlue50,
                IldsTokens.informativeBlue500,
                Icons.Filled.Info,
            )
            IldsToastVariant.Warning -> IldsToastStyle(
                IldsTokens.warningAmber50,
                IldsTokens.warningAmber500,
                Icons.Filled.Warning,
            )
            IldsToastVariant.Error -> IldsToastStyle(
                IldsTokens.errorRed50,
                IldsTokens.errorRed600,
                Icons.Filled.Error,
            )
        }
    }
}
